File: benchmarks/BitArray/c/bitarray_bench.c
// Copyright (c) 2026, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.
//
// C reference benchmark for BitArray cardinality, mirroring the Dart
// version in ../dart/BitArray.dart. Uses __builtin_popcount as the
// "intrinsic" path and a hand-written SWAR Hamming weight as the
// software path, with both un-unrolled and 8-way unrolled variants.
//
// Compile: cc -O3 -std=c11 bitarray_bench.c -o bitarray_bench
// Run:     ./bitarray_bench

#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#define BENCH_BITS (1u << 20)
#define WORD_BITS  32
#define WORD_SHIFT 5
#define WORD_MASK  (WORD_BITS - 1)
#define N_WORDS    ((BENCH_BITS + WORD_BITS - 1) >> WORD_SHIFT)

// Match BenchmarkBase's reporting unit: microseconds per single run() call.
// To pick an iteration count, we measure once at ~50 ms and scale.

static uint32_t bits[N_WORDS];

static void fill_quarter_density(uint32_t* w, size_t n) {
  // Deterministic xorshift32 with a fixed seed, then keep only ~25% of bits.
  uint32_t s = 0xBEEFu;
  for (size_t i = 0; i < n; i++) {
    // 4 random uint32s combined: AND any 2 of 4 -> ~25% bit density.
    uint32_t r1, r2;
    s ^= s << 13; s ^= s >> 17; s ^= s << 5; r1 = s;
    s ^= s << 13; s ^= s >> 17; s ^= s << 5; r2 = s;
    w[i] = r1 & r2;
  }
}

// ----- cardinality kernels --------------------------------------------------

static int cardinality_intrinsic(const uint32_t* w, size_t n) {
  int total = 0;
  for (size_t i = 0; i < n; i++) {
    total += __builtin_popcount(w[i]);
  }
  return total;
}

static int cardinality_intrinsic_unrolled(const uint32_t* w, size_t n) {
  size_t limit = n - (n & 7);
  int total = 0;
  for (size_t i = 0; i < limit; i += 8) {
    total +=
        __builtin_popcount(w[i]) +
        __builtin_popcount(w[i + 1]) +
        __builtin_popcount(w[i + 2]) +
        __builtin_popcount(w[i + 3]) +
        __builtin_popcount(w[i + 4]) +
        __builtin_popcount(w[i + 5]) +
        __builtin_popcount(w[i + 6]) +
        __builtin_popcount(w[i + 7]);
  }
  for (size_t i = limit; i < n; i++) {
    total += __builtin_popcount(w[i]);
  }
  return total;
}

static int cardinality_swar(const uint32_t* w, size_t n) {
  int total = 0;
  for (size_t i = 0; i < n; i++) {
    uint32_t v = w[i];
    v = v - ((v >> 1) & 0x55555555u);
    v = (v & 0x33333333u) + ((v >> 2) & 0x33333333u);
    v = (v + (v >> 4)) & 0x0F0F0F0Fu;
    total += (int)((v * 0x01010101u) >> 24);
  }
  return total;
}

static int cardinality_swar_unrolled(const uint32_t* w, size_t n) {
  size_t limit = n - (n & 7);
  int total = 0;
  for (size_t i = 0; i < limit; i += 8) {
    uint32_t a = w[i], b = w[i + 1], c = w[i + 2], d = w[i + 3];
    uint32_t e = w[i + 4], f = w[i + 5], g = w[i + 6], h = w[i + 7];
    a = a - ((a >> 1) & 0x55555555u);
    b = b - ((b >> 1) & 0x55555555u);
    c = c - ((c >> 1) & 0x55555555u);
    d = d - ((d >> 1) & 0x55555555u);
    e = e - ((e >> 1) & 0x55555555u);
    f = f - ((f >> 1) & 0x55555555u);
    g = g - ((g >> 1) & 0x55555555u);
    h = h - ((h >> 1) & 0x55555555u);
    a = (a & 0x33333333u) + ((a >> 2) & 0x33333333u);
    b = (b & 0x33333333u) + ((b >> 2) & 0x33333333u);
    c = (c & 0x33333333u) + ((c >> 2) & 0x33333333u);
    d = (d & 0x33333333u) + ((d >> 2) & 0x33333333u);
    e = (e & 0x33333333u) + ((e >> 2) & 0x33333333u);
    f = (f & 0x33333333u) + ((f >> 2) & 0x33333333u);
    g = (g & 0x33333333u) + ((g >> 2) & 0x33333333u);
    h = (h & 0x33333333u) + ((h >> 2) & 0x33333333u);
    a = (a + (a >> 4)) & 0x0F0F0F0Fu;
    b = (b + (b >> 4)) & 0x0F0F0F0Fu;
    c = (c + (c >> 4)) & 0x0F0F0F0Fu;
    d = (d + (d >> 4)) & 0x0F0F0F0Fu;
    e = (e + (e >> 4)) & 0x0F0F0F0Fu;
    f = (f + (f >> 4)) & 0x0F0F0F0Fu;
    g = (g + (g >> 4)) & 0x0F0F0F0Fu;
    h = (h + (h >> 4)) & 0x0F0F0F0Fu;
    total +=
        (int)((a * 0x01010101u) >> 24) +
        (int)((b * 0x01010101u) >> 24) +
        (int)((c * 0x01010101u) >> 24) +
        (int)((d * 0x01010101u) >> 24) +
        (int)((e * 0x01010101u) >> 24) +
        (int)((f * 0x01010101u) >> 24) +
        (int)((g * 0x01010101u) >> 24) +
        (int)((h * 0x01010101u) >> 24);
  }
  for (size_t i = limit; i < n; i++) {
    uint32_t v = w[i];
    v = v - ((v >> 1) & 0x55555555u);
    v = (v & 0x33333333u) + ((v >> 2) & 0x33333333u);
    v = (v + (v >> 4)) & 0x0F0F0F0Fu;
    total += (int)((v * 0x01010101u) >> 24);
  }
  return total;
}

// ----- harness --------------------------------------------------------------

static double now_us(void) {
  struct timespec ts;
  clock_gettime(CLOCK_MONOTONIC, &ts);
  return ts.tv_sec * 1e6 + ts.tv_nsec / 1e3;
}

static volatile int sink;

static void bench(const char* name, int (*fn)(const uint32_t*, size_t)) {
  // Warmup.
  for (int i = 0; i < 16; i++) sink ^= fn(bits, N_WORDS);

  // Calibrate iteration count to roughly 100 ms.
  double t0 = now_us();
  int iters = 0;
  int target_us = 100000;
  while (now_us() - t0 < target_us) {
    sink ^= fn(bits, N_WORDS);
    iters++;
  }
  double elapsed = now_us() - t0;
  double us_per_iter = elapsed / iters;
  printf("BitArray.%s(RunTime): %.4f us.\n", name, us_per_iter);
}

int main(void) {
  fill_quarter_density(bits, N_WORDS);

  // Sanity: all four kernels must agree.
  int c1 = cardinality_intrinsic(bits, N_WORDS);
  int c2 = cardinality_intrinsic_unrolled(bits, N_WORDS);
  int c3 = cardinality_swar(bits, N_WORDS);
  int c4 = cardinality_swar_unrolled(bits, N_WORDS);
  if (c1 != c2 || c1 != c3 || c1 != c4) {
    fprintf(stderr, "FAIL: %d %d %d %d\n", c1, c2, c3, c4);
    return 1;
  }
  fprintf(stderr, "cardinality (sanity check) = %d / %u\n", c1, BENCH_BITS);

  bench("cardinality.swar", cardinality_swar);
  bench("cardinality.swar.unrolled", cardinality_swar_unrolled);
  bench("cardinality.intrinsic", cardinality_intrinsic);
  bench("cardinality.intrinsic.unrolled", cardinality_intrinsic_unrolled);

  return 0;
}

File: benchmarks/BitArray/rust/bitarray_bench.rs
// Copyright (c) 2026, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.
//
// Rust reference benchmark for BitArray cardinality, mirroring the Dart
// version in ../dart/BitArray.dart and the C version in
// ../c/bitarray_bench.c. Uses u32::count_ones as the "intrinsic" path
// and a hand-written SWAR Hamming weight as the software path, with
// both un-unrolled and 8-way unrolled variants.
//
// Build: rustc -C opt-level=3 bitarray_bench.rs -o bitarray_bench_rs
// Run:   ./bitarray_bench_rs

use std::hint::black_box;
use std::time::Instant;

const BENCH_BITS: usize = 1 << 20;
const WORD_BITS: usize = 32;
const N_WORDS: usize = (BENCH_BITS + WORD_BITS - 1) / WORD_BITS;

fn fill_quarter_density(w: &mut [u32]) {
    // Match C's xorshift32 + AND-of-two so the bit pattern is identical.
    let mut s: u32 = 0xBEEF;
    for slot in w.iter_mut() {
        s ^= s << 13;
        s ^= s >> 17;
        s ^= s << 5;
        let r1 = s;
        s ^= s << 13;
        s ^= s >> 17;
        s ^= s << 5;
        let r2 = s;
        *slot = r1 & r2;
    }
}

// ----- cardinality kernels --------------------------------------------------

#[inline(never)]
fn cardinality_intrinsic(w: &[u32]) -> u32 {
    let mut total = 0u32;
    for &v in w {
        total += v.count_ones();
    }
    total
}

#[inline(never)]
fn cardinality_intrinsic_unrolled(w: &[u32]) -> u32 {
    let n = w.len();
    let limit = n - (n & 7);
    let mut total = 0u32;
    let mut i = 0;
    while i < limit {
        total += w[i].count_ones()
            + w[i + 1].count_ones()
            + w[i + 2].count_ones()
            + w[i + 3].count_ones()
            + w[i + 4].count_ones()
            + w[i + 5].count_ones()
            + w[i + 6].count_ones()
            + w[i + 7].count_ones();
        i += 8;
    }
    while i < n {
        total += w[i].count_ones();
        i += 1;
    }
    total
}

#[inline(never)]
fn cardinality_swar(w: &[u32]) -> u32 {
    let mut total = 0u32;
    for &word in w {
        let mut v = word;
        v = v.wrapping_sub((v >> 1) & 0x55555555);
        v = (v & 0x33333333).wrapping_add((v >> 2) & 0x33333333);
        v = v.wrapping_add(v >> 4) & 0x0F0F0F0F;
        total += v.wrapping_mul(0x01010101) >> 24;
    }
    total
}

#[inline(never)]
fn cardinality_swar_unrolled(w: &[u32]) -> u32 {
    let n = w.len();
    let limit = n - (n & 7);
    let mut total = 0u32;
    let mut i = 0;
    while i < limit {
        let mut a = w[i];
        let mut b = w[i + 1];
        let mut c = w[i + 2];
        let mut d = w[i + 3];
        let mut e = w[i + 4];
        let mut f = w[i + 5];
        let mut g = w[i + 6];
        let mut h = w[i + 7];
        a = a.wrapping_sub((a >> 1) & 0x55555555);
        b = b.wrapping_sub((b >> 1) & 0x55555555);
        c = c.wrapping_sub((c >> 1) & 0x55555555);
        d = d.wrapping_sub((d >> 1) & 0x55555555);
        e = e.wrapping_sub((e >> 1) & 0x55555555);
        f = f.wrapping_sub((f >> 1) & 0x55555555);
        g = g.wrapping_sub((g >> 1) & 0x55555555);
        h = h.wrapping_sub((h >> 1) & 0x55555555);
        a = (a & 0x33333333).wrapping_add((a >> 2) & 0x33333333);
        b = (b & 0x33333333).wrapping_add((b >> 2) & 0x33333333);
        c = (c & 0x33333333).wrapping_add((c >> 2) & 0x33333333);
        d = (d & 0x33333333).wrapping_add((d >> 2) & 0x33333333);
        e = (e & 0x33333333).wrapping_add((e >> 2) & 0x33333333);
        f = (f & 0x33333333).wrapping_add((f >> 2) & 0x33333333);
        g = (g & 0x33333333).wrapping_add((g >> 2) & 0x33333333);
        h = (h & 0x33333333).wrapping_add((h >> 2) & 0x33333333);
        a = a.wrapping_add(a >> 4) & 0x0F0F0F0F;
        b = b.wrapping_add(b >> 4) & 0x0F0F0F0F;
        c = c.wrapping_add(c >> 4) & 0x0F0F0F0F;
        d = d.wrapping_add(d >> 4) & 0x0F0F0F0F;
        e = e.wrapping_add(e >> 4) & 0x0F0F0F0F;
        f = f.wrapping_add(f >> 4) & 0x0F0F0F0F;
        g = g.wrapping_add(g >> 4) & 0x0F0F0F0F;
        h = h.wrapping_add(h >> 4) & 0x0F0F0F0F;
        total += a.wrapping_mul(0x01010101) >> 24;
        total += b.wrapping_mul(0x01010101) >> 24;
        total += c.wrapping_mul(0x01010101) >> 24;
        total += d.wrapping_mul(0x01010101) >> 24;
        total += e.wrapping_mul(0x01010101) >> 24;
        total += f.wrapping_mul(0x01010101) >> 24;
        total += g.wrapping_mul(0x01010101) >> 24;
        total += h.wrapping_mul(0x01010101) >> 24;
        i += 8;
    }
    while i < n {
        let mut v = w[i];
        v = v.wrapping_sub((v >> 1) & 0x55555555);
        v = (v & 0x33333333).wrapping_add((v >> 2) & 0x33333333);
        v = v.wrapping_add(v >> 4) & 0x0F0F0F0F;
        total += v.wrapping_mul(0x01010101) >> 24;
        i += 1;
    }
    total
}

// ----- harness --------------------------------------------------------------

fn bench(name: &str, fn_: impl Fn(&[u32]) -> u32, w: &[u32]) {
    // Warmup.
    for _ in 0..16 {
        black_box(fn_(black_box(w)));
    }
    // Calibrate to roughly 100 ms wall.
    let target_us = 100_000u128;
    let t0 = Instant::now();
    let mut iters = 0u32;
    while (t0.elapsed().as_micros()) < target_us {
        black_box(fn_(black_box(w)));
        iters += 1;
    }
    let elapsed_us = t0.elapsed().as_micros() as f64;
    let us_per_iter = elapsed_us / (iters as f64);
    println!("BitArray.{}(RunTime): {:.4} us.", name, us_per_iter);
}

fn main() {
    let mut bits = vec![0u32; N_WORDS];
    fill_quarter_density(&mut bits);

    // Sanity: all four kernels must agree.
    let c1 = cardinality_intrinsic(&bits);
    let c2 = cardinality_intrinsic_unrolled(&bits);
    let c3 = cardinality_swar(&bits);
    let c4 = cardinality_swar_unrolled(&bits);
    assert_eq!(c1, c2);
    assert_eq!(c1, c3);
    assert_eq!(c1, c4);
    eprintln!("cardinality (sanity check) = {} / {}", c1, BENCH_BITS);

    bench("cardinality.swar", cardinality_swar, &bits);
    bench("cardinality.swar.unrolled", cardinality_swar_unrolled, &bits);
    bench("cardinality.intrinsic", cardinality_intrinsic, &bits);
    bench("cardinality.intrinsic.unrolled", cardinality_intrinsic_unrolled, &bits);
}
